Docs 7260 Kafka 4.0 (#652)
mulesoft/docs-connectors: modules/ROOT/pages/kafka/kafka-connector-upgrade-migrate.adoc

= Kafka - Upgrade and Migrate - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

Upgrade Anypoint Connector for Kafka (Kafka Connector) to version 4.0.0.

== Supported Upgrade Paths

[%header,cols="50a,50a"]
|===
|From Version | To Version
|3.2.x |4.0.x
|===

== Changes in This Release

This release contains the following changes:

> * New operations with new parameters +

Reviewer: ```suggestion
* New operations with new parameters were added.+
```

See the <<new_operations,New Operations>> table.
* The Message Listener replaces the Message Consumer source. +
See the <<changed_sources,Changed Sources>> table.

> * A new source was added +

Reviewer: ```suggestion
* A new source was added. +
```

See the <<new_sources,New Sources>> table for information about the new parameters.
* Added different acknowledgment modes in the consumer configuration, including:
** AUTO +
Mule acknowledges the message only if the flow is finished successfully. 
** MANUAL +
The user must manually acknowledge the message within the flow.
** DUPS_OK + 
Same as the AUTO mode, but the commit is made asynchronously, which can result in duplicate records.
** IMMEDIATE +
Mule automatically acknowledges the message upon receipt.

> * Added a Transactional SASL connection +

Reviewer: ```suggestion
* Added a Transactional SASL connection. +
```

This connection type is the same as the Producer SASL connection, except that it’s not a cached connection and is transactional.

> * Added a Transactional plaintext connection +

Reviewer: ```suggestion
* Added a Transactional plaintext connection. +
```

This connection type is the same as the Producer plaintext connection, except that it’s not a cached connection and is transactional.

[[changed_sources]]
== Changed Sources

[%header%autowidth.spread]
|===

> |Kafka Source | Old version | Parameters

Reviewer: ```suggestion
|Kafka Source | Former Version | Parameters
```

> | Message listener | Message consumer a| * Poll timeout

Reviewer: Shouldn't this be Message Listener?

* Poll timeout time unit
* Acknowledgment mode
* Number of parallel consumers
|===

[[new_operations]]
== New Operations

[%header%autowidth.spread]
|===
|Kafka Operation | Description | Parameters

| Commit

> a| Commits the offsets associated to a message or batch of messages consumed in a message listener. a| Consumer commit key. The consumer commit key to use to commit the message.

Reviewer: Message Listener?

> | Consume | This operation enables receiving messages from one or more Kafka topics. It works very similarly to the Message Listener source, so all the operations that apply to the Message Listener source also apply to this operation. a| * Consumption timeout

Reviewer: ```suggestion
| Consume | Enables receiving messages from one or more Kafka topics. It works very similarly to the Message Listener source, so all the operations that apply to the Message Listener source also apply to this operation. a| * Consumption timeout
```

* Timeout time unit

> |Describe | Retrieves the subscriptions and assignments of the associated connection, along with the current offsets for each topic-partition. This operation blocks all the consumers, so there are no changes to the offsets assignments or other information that is retrieved by this operation. |

Reviewer: ```suggestion
|Describe | Retrieves the subscriptions and assignments of the associated connection, along with the current offsets for each topic partition. This operation blocks all the consumers, so there are no changes to the offsets assignments or other information that is retrieved by this operation. |
```

> | Publish | The `publish` operation enables sending messages to a Kafka topic, and is non-blocking. Its behavior depends on specific values of the connection parameters. This operation supports transactions, which means that a transaction ID is randomly generated for the Kafka producer and handled in the connection.

Reviewer: ```suggestion
| Publish | Enables sending messages to a Kafka topic, and is nonblocking. Its behavior depends on specific values of the connection parameters. This operation supports transactions, which means that a transaction ID is randomly generated for the Kafka producer and handled in the connection.
```

a| *Topic (The topic name)
* Partition (The partition name)
* Key
* Message
* Headers
| Seek | Sets the current offset of the consumer for the given topic and partition to the provided offset value. a| * Topic (The topic name)
* Partition (The partition name)
* Offset (The offset to seek to) |
|===

[[new_sources]]
== New Sources

[%header%autowidth.spread]
|===
|Kafka Operation | Description | Parameters

> | Batch message listener

Reviewer: Doublecheck capitalization of message listener throughout

a| Works in the same way as the message listener except that it processes the list of messages instead of processing a single message at a time. The message list that was obtained in the poll is handled by a flow as a single event, so the handling of concurrency is simpler than in the simple message listener. In other words, a commit of the messages happens for all the messages as a whole by calling the commit on Kafka consumer.
a| * Poll timeout
* Poll timeout time unit
* Acknowledgment mode
* Number of parallel consumers
|===

== Upgrade Prerequisites

Before you perform the upgrade, you must create a backup of your files, data, and configurations in case you need to revert to the previous version.

== Upgrade Steps

Follow these steps to perform the upgrade to Kafka Connector 4.0.x:

. In Studio, create a Mule project.
. In the Mule Palette view, click *Search in Exchange*.
. In Add Modules to Project, type 'Kafka' in the search field.

> . In Available modules, select *Kafka Connector*, and then click *Add*.

Reviewer: ```suggestion
. In Available modules, select *Kafka Connector* and then click *Add*.
```

. Click *Finish*. +
Anypoint Studio upgrades the connector automatically.

> . Verify that the `kafka-connector` dependency version is `4.0.x` in the pom.xml.

Reviewer: This sentence is formatted differently throughout this topic set: please search for and standardize.

== Post Upgrade Steps

After you install the latest version of the connector, follow these steps to complete the upgrade:

. In Anypoint Studio, verify that there are no errors in the *Problems* or *Console* views.
. Check the project pom.xml and verify that there are no problems.
. Test the connection and verify that the operations work.

== Troubleshooting

If there are problems with caching the parameters and caching the metadata, try restarting Anypoint Studio.

=== Reverting the Upgrade

> If it is necessary to revert to the previous version of Kafka Connector, in Anypoint Studio, change the `kafka-connector` dependency version `4.0.x` in the project's pom.xml to the previous version.

Reviewer: ditto
Diane formats this differently as well. You two should prolly coordinate at some point. I know you are in a rush right now.

= Kafka - Upgrade and Migrate - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

Upgrade Anypoint Connector for Kafka (Kafka Connector) to version 4.0.0.

== Supported Upgrade Paths

[%header,cols="50a,50a"]
|===
|From Version | To Version
|3.2.x |4.0.x
|===

== Changes in This Release

Kafka Connector 4.0.0 contains a lot of changes to improve efficiency, including new operations to enable you to better understand how to use the connector.

All of the changes included in this version break backward compatibility.

This release contains the following changes:

* Five new operations with new parameters were added.+
 ** Commit + 
 This operation commits the offsets associated with a message or batch of messages consumed in a Message Listener (Source) or in a Batch Message Listener (Source). This operation works only with MANUAL Ack mode.
 ** Consume +
 This operation is used for receiving messages from one or more Kafka topics. The Consume operation works similarly to the Message Listener source, so all the operations that apply to the Message Listener source also apply to this operation.
 ** Describe +

>  This operation is used for retrieving the subscriptions and assignments of the associated connection, along with the current offsets for each topic-partition. This operation blocks all the Consumers (so there are no changes to the offsets assignments or other information) that are retrieved by this operation. The Describe operation returns a list of topics with their partition assignments and their offsets.

Reviewer 2: Describe operation was removed for release 4.0.0

** DUPS_OK + 
Same as the AUTO mode, but the commit is made asynchronously, which can result in duplicate records.
** IMMEDIATE +
Mule automatically acknowledges the message upon receipt.
* The offset paramater was added in the Consumer configuration.
* There is now one plaintext connection type that supports transactions. 

[[changed_sources]]
== Changed Sources

[%header%autowidth.spread]
|===
|Kafka Source | Earlier version | Parameters
| Message Listener | Message consumer a| * Poll timeout
* Poll timeout time unit
* Acknowledgment mode
* Number of parallel consumers
|===

[[new_operations]]
== New Operations

[%header%autowidth.spread]
|===
|Kafka Operation | Description | Parameters
| Commit
a| Commits the offsets associated to a message or batch of messages consumed in a Message Listener. a| Consumer commit key. The consumer commit key to use to commit the message.
| Consume | Enables receiving messages from one or more Kafka topics. It works very similarly to the Message Listener source, so all the operations that apply to the Message Listener source also apply to this operation. a| * Consumption timeout
* Timeout time unit

> |Describe | Retrieves the subscriptions and assignments of the associated connection, along with the current offsets for each topic partition. This operation blocks all the consumers, so there are no changes to the offsets assignments or other information that is retrieved by this operation. |

Reviewer 2: Describe was removed for this release (4.0.0)
